File: tools/testing/connection/undici/elasticsearch.mjs
/**
 * Elasticsearch - undici Connection Test
 *
 * Authentication: Basic (username:password)
 * Base URL: http://localhost:9200 or https://...
 * Health Endpoint: GET /_cluster/health
 *
 * TLS/SSL Options:
 *   NODE_TLS_REJECT_UNAUTHORIZED=0  - Ignore all certificate errors
 *   REQUEST_CA_BUNDLE=/path/to/ca   - Custom CA bundle file
 *   SSL_CERT_FILE=/path/to/cert     - Custom SSL certificate file
 *   NODE_EXTRA_CA_CERTS=/path/to/ca - Additional CA certificates
 */

import { request, ProxyAgent, Agent } from 'undici';
import { readFileSync } from 'node:fs';

// ============================================================================
// Configuration - Override these values
// ============================================================================

const CONFIG = {
  // Option 1: Full Connection URL
  ELASTIC_DB_URL: process.env.ELASTIC_DB_URL || '',

  // Option 2: Individual Components
  ELASTIC_DB_HOST: process.env.ELASTIC_DB_HOST || 'localhost',
  ELASTIC_DB_PORT: process.env.ELASTIC_DB_PORT || '9200',
  ELASTIC_DB_USERNAME: process.env.ELASTIC_DB_USERNAME || '',
  ELASTIC_DB_ACCESS_KEY: process.env.ELASTIC_DB_ACCESS_KEY || '',
  ELASTIC_DB_TLS: process.env.ELASTIC_DB_TLS || 'false', // 'true' for HTTPS

  // Optional: Proxy Configuration
  HTTPS_PROXY: process.env.HTTPS_PROXY || '', // e.g., 'http://proxy.example.com:8080'

  // Optional: TLS Configuration
  // Set NODE_TLS_REJECT_UNAUTHORIZED=0 to ignore certificate errors
  REJECT_UNAUTHORIZED: process.env.NODE_TLS_REJECT_UNAUTHORIZED !== '0',

  // Optional: Custom CA certificates
  // REQUEST_CA_BUNDLE or SSL_CERT_FILE - path to custom CA bundle
  // NODE_EXTRA_CA_CERTS - handled automatically by Node.js
  CA_BUNDLE: process.env.REQUEST_CA_BUNDLE || process.env.SSL_CERT_FILE || '',
};

// ============================================================================
// TLS Configuration Helper
// ============================================================================

function getTlsConnectOptions() {
  const options = {
    rejectUnauthorized: CONFIG.REJECT_UNAUTHORIZED,
  };

  // Load custom CA bundle if specified
  if (CONFIG.CA_BUNDLE) {
    try {
      options.ca = readFileSync(CONFIG.CA_BUNDLE);
      console.log(`Using custom CA bundle: ${CONFIG.CA_BUNDLE}`);
    } catch (err) {
      console.warn(`Warning: Could not read CA bundle: ${err.message}`);
    }
  }

  return options;
}

// ============================================================================
// Build Base URL
// ============================================================================

function getBaseUrl() {
  if (CONFIG.ELASTIC_DB_URL) {
    return CONFIG.ELASTIC_DB_URL;
  }

  const useTls = CONFIG.ELASTIC_DB_TLS === 'true' ||
    ['443', '9243', '25060'].includes(CONFIG.ELASTIC_DB_PORT);
  const protocol = useTls ? 'https' : 'http';

  let url = `${protocol}://`;

  if (CONFIG.ELASTIC_DB_USERNAME && CONFIG.ELASTIC_DB_ACCESS_KEY) {
    url += `${CONFIG.ELASTIC_DB_USERNAME}:${CONFIG.ELASTIC_DB_ACCESS_KEY}@`;
  }

  url += `${CONFIG.ELASTIC_DB_HOST}:${CONFIG.ELASTIC_DB_PORT}`;

  return url;
}

// ============================================================================
// Create Dispatcher (with or without proxy)
// ============================================================================

function createDispatcher() {
  const tlsOptions = getTlsConnectOptions();

  if (CONFIG.HTTPS_PROXY) {
    console.log(`Using proxy: ${CONFIG.HTTPS_PROXY}`);
    return new ProxyAgent({
      uri: CONFIG.HTTPS_PROXY,
      connect: tlsOptions,
    });
  }
  return new Agent({
    connect: tlsOptions,
  });
}

// ============================================================================
// Create Basic Auth Header
// ============================================================================

function createBasicAuthHeader() {
  if (!CONFIG.ELASTIC_DB_USERNAME || !CONFIG.ELASTIC_DB_ACCESS_KEY) {
    return null;
  }
  const credentials = `${CONFIG.ELASTIC_DB_USERNAME}:${CONFIG.ELASTIC_DB_ACCESS_KEY}`;
  return `Basic ${Buffer.from(credentials).toString('base64')}`;
}

// ============================================================================
// Health Check
// ============================================================================

async function healthCheck() {
  console.log('\n=== Elasticsearch Health Check ===\n');

  const dispatcher = createDispatcher();
  const baseUrl = getBaseUrl();

  try {
    const headers = {
      'Accept': 'application/json',
    };

    const authHeader = createBasicAuthHeader();
    if (authHeader) {
      headers['Authorization'] = authHeader;
    }

    const response = await request(`${baseUrl}/_cluster/health`, {
      method: 'GET',
      headers,
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

// ============================================================================
// Sample API Calls
// ============================================================================

async function getClusterInfo() {
  console.log('\n=== Cluster Info ===\n');

  const dispatcher = createDispatcher();
  const baseUrl = getBaseUrl();

  try {
    const headers = { 'Accept': 'application/json' };
    const authHeader = createBasicAuthHeader();
    if (authHeader) headers['Authorization'] = authHeader;

    const response = await request(baseUrl, {
      method: 'GET',
      headers,
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function listIndices() {
  console.log('\n=== List Indices ===\n');

  const dispatcher = createDispatcher();
  const baseUrl = getBaseUrl();

  try {
    const headers = { 'Accept': 'application/json' };
    const authHeader = createBasicAuthHeader();
    if (authHeader) headers['Authorization'] = authHeader;

    const response = await request(`${baseUrl}/_cat/indices?format=json`, {
      method: 'GET',
      headers,
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log(`Found ${data.length} indices`);
    data.slice(0, 10).forEach((index) => {
      console.log(`  - ${index.index} (${index.health}, ${index['docs.count']} docs)`);
    });

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function searchIndex(indexName, query = { match_all: {} }) {
  console.log(`\n=== Search Index: ${indexName} ===\n`);

  const dispatcher = createDispatcher();
  const baseUrl = getBaseUrl();

  try {
    const headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
    };
    const authHeader = createBasicAuthHeader();
    if (authHeader) headers['Authorization'] = authHeader;

    const response = await request(`${baseUrl}/${indexName}/_search`, {
      method: 'POST',
      headers,
      body: JSON.stringify({ query }),
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log(`Found ${data.hits?.total?.value || 0} documents`);

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function indexDocument(indexName, document) {
  console.log(`\n=== Index Document: ${indexName} ===\n`);

  const dispatcher = createDispatcher();
  const baseUrl = getBaseUrl();

  try {
    const headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
    };
    const authHeader = createBasicAuthHeader();
    if (authHeader) headers['Authorization'] = authHeader;

    const response = await request(`${baseUrl}/${indexName}/_doc`, {
      method: 'POST',
      headers,
      body: JSON.stringify(document),
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 201, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

// ============================================================================
// Run Tests
// ============================================================================

async function main() {
  const baseUrl = getBaseUrl();

  console.log('Elasticsearch Connection Test');
  console.log('=============================');
  console.log(`Base URL: ${baseUrl.replace(/:[^:@]+@/, ':***@')}`); // Mask password
  console.log(`Proxy: ${CONFIG.HTTPS_PROXY || 'None'}`);
  console.log(`TLS Verify: ${CONFIG.REJECT_UNAUTHORIZED}`);
  console.log(`CA Bundle: ${CONFIG.CA_BUNDLE || 'System default'}`);

  await healthCheck();
  // await getClusterInfo();
  // await listIndices();
  // await searchIndex('my-index');
  // await indexDocument('my-index', { title: 'Test', content: 'Hello World' });
}

main().catch(console.error);

File: tools/testing/connection/httpx/elasticsearch.py
#!/usr/bin/env python3
"""
Elasticsearch API - httpx Connection Test

Authentication: Basic (username:password)
Base URL: https://{host}:9200
Health Endpoint: GET /_cluster/health

TLS/SSL Options:
  SSL_CERT_VERIFY=0              - Ignore all certificate errors
  REQUEST_CA_BUNDLE=/path/to/ca  - Custom CA bundle file
  SSL_CERT_FILE=/path/to/cert    - Custom SSL certificate file
  REQUESTS_CA_BUNDLE=/path/to/ca - Alternative CA bundle (requests compat)
"""

import asyncio
import base64
import json
import os
from typing import Any, Union

import httpx

# ============================================================================
# Configuration - Override these values
# ============================================================================

CONFIG = {
    # Required
    "ELASTICSEARCH_HOST": os.getenv("ELASTICSEARCH_HOST", "localhost"),
    "ELASTICSEARCH_PORT": os.getenv("ELASTICSEARCH_PORT", "9200"),
    "ELASTICSEARCH_USER": os.getenv("ELASTICSEARCH_USER", "elastic"),
    "ELASTICSEARCH_PASSWORD": os.getenv("ELASTICSEARCH_PASSWORD", "your_password_here"),

    # Optional: Use HTTPS
    "ELASTICSEARCH_USE_SSL": os.getenv("ELASTICSEARCH_USE_SSL", "true").lower() == "true",

    # Optional: Proxy Configuration
    "HTTPS_PROXY": os.getenv("HTTPS_PROXY", ""),  # e.g., "http://proxy.example.com:8080"

    # Optional: TLS Configuration
    # Set SSL_CERT_VERIFY=0 to ignore certificate errors (default for Elasticsearch)
    "VERIFY_SSL": os.getenv("SSL_CERT_VERIFY", "0") != "0",  # Elasticsearch often uses self-signed certs

    # Optional: Custom CA certificates
    # REQUEST_CA_BUNDLE, SSL_CERT_FILE, or REQUESTS_CA_BUNDLE - path to custom CA bundle
    "CA_BUNDLE": os.getenv("REQUEST_CA_BUNDLE") or os.getenv("SSL_CERT_FILE") or os.getenv("REQUESTS_CA_BUNDLE") or "",
}

# Build base URL
PROTOCOL = "https" if CONFIG["ELASTICSEARCH_USE_SSL"] else "http"
BASE_URL = f"{PROTOCOL}://{CONFIG['ELASTICSEARCH_HOST']}:{CONFIG['ELASTICSEARCH_PORT']}"


# ============================================================================
# TLS Configuration Helper
# ============================================================================

def get_ssl_verify() -> Union[bool, str]:
    """Get SSL verification setting."""
    if CONFIG["CA_BUNDLE"]:
        print(f"Using custom CA bundle: {CONFIG['CA_BUNDLE']}")
        return CONFIG["CA_BUNDLE"]
    return CONFIG["VERIFY_SSL"]


# ============================================================================
# Create HTTP Client
# ============================================================================

def create_client() -> httpx.AsyncClient:
    """Create httpx async client with optional proxy."""
    proxies = None
    if CONFIG["HTTPS_PROXY"]:
        print(f"Using proxy: {CONFIG['HTTPS_PROXY']}")
        proxies = {
            "http://": CONFIG["HTTPS_PROXY"],
            "https://": CONFIG["HTTPS_PROXY"],
        }

    return httpx.AsyncClient(
        proxies=proxies,
        verify=get_ssl_verify(),
        timeout=httpx.Timeout(30.0),
    )


def create_basic_auth_header() -> str:
    """Create Basic auth header."""
    credentials = f"{CONFIG['ELASTICSEARCH_USER']}:{CONFIG['ELASTICSEARCH_PASSWORD']}"
    encoded = base64.b64encode(credentials.encode()).decode()
    return f"Basic {encoded}"


# ============================================================================
# Health Check
# ============================================================================

async def health_check() -> dict[str, Any]:
    """Perform health check against Elasticsearch."""
    print("\n=== Elasticsearch Health Check ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{BASE_URL}/_cluster/health",
                headers={
                    "Authorization": create_basic_auth_header(),
                    "Accept": "application/json",
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"Cluster: {data.get('cluster_name')}")
            print(f"Health: {data.get('status')}")
            print(f"Nodes: {data.get('number_of_nodes')}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Sample API Calls
# ============================================================================

async def get_cluster_info() -> dict[str, Any]:
    """Get cluster information."""
    print("\n=== Get Cluster Info ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                BASE_URL,
                headers={
                    "Authorization": create_basic_auth_header(),
                    "Accept": "application/json",
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"Response: {json.dumps(data, indent=2)}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def list_indices() -> dict[str, Any]:
    """List all indices."""
    print("\n=== List Indices ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{BASE_URL}/_cat/indices",
                params={"format": "json"},
                headers={
                    "Authorization": create_basic_auth_header(),
                    "Accept": "application/json",
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"Found {len(data)} indices")
            for index in data[:10]:
                print(f"  - {index['index']} ({index.get('docs.count', 0)} docs)")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def search_index(index: str, query: dict) -> dict[str, Any]:
    """Search an index."""
    print(f"\n=== Search Index: {index} ===\n")

    async with create_client() as client:
        try:
            response = await client.post(
                f"{BASE_URL}/{index}/_search",
                headers={
                    "Authorization": create_basic_auth_header(),
                    "Accept": "application/json",
                    "Content-Type": "application/json",
                },
                json=query,
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            hits = data.get("hits", {})
            print(f"Total hits: {hits.get('total', {}).get('value', 0)}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Run Tests
# ============================================================================

async def main():
    """Run connection tests."""
    print("Elasticsearch API Connection Test")
    print("==================================")
    print(f"Base URL: {BASE_URL}")
    print(f"Proxy: {CONFIG['HTTPS_PROXY'] or 'None'}")
    print(f"User: {CONFIG['ELASTICSEARCH_USER']}")
    print(f"SSL Verify: {CONFIG['VERIFY_SSL']}")
    print(f"CA Bundle: {CONFIG['CA_BUNDLE'] or 'System default'}")

    await health_check()
    # await get_cluster_info()
    # await list_indices()
    # await search_index("my-index", {"query": {"match_all": {}}})


if __name__ == "__main__":
    asyncio.run(main())
